File: 137-single-number-ii/solution.ts
// https://leetcode.com/problems/single-number-ii/solutions/3621277/simple-solution-with-o-1-space-o-n-time-complexity:
function singleNumber(nums: number[]): number {
  let answer = 0;
  for (let i = 0; i < 32; i++) {
    let ones = 0;
    let zeroes = 0;
    for (let n of nums) (n >> i) & 1 ? ones++ : zeroes++;
    if (ones % 3) answer |= 1 << i;
  }
  return answer;
}

File: 137-single-number-ii/solution.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int ans = 0, b = 1;
        while(true) {
            int cnt = 0;
            for(auto& num : nums) cnt += ((num & b) != 0);
            if(cnt % 3 != 0) ans |= b;
            if(b == INT_MIN) break;
            b <<= 1;
        }
        return ans;
    }
};
